text/3389-manifest-lint.md

```diff
@@ -0,0 +1,578 @@
+[lints.rust]
+unsafe = { level = "forbid", priority = 0 }
+```
+
+`level` corresponds to the lint levels in `rustc`:
+- `forbid`
+- `deny`
+- `warn`
+- `allow`
+
+`priority` is a signed integer that controls which lints or lint groups override other lint groups:
+- lower (particularly negative) numbers have lower priority, being overridden
+  by higher numbers, and show up first on the command-line to tools like
+  `rustc`
+
+To know which table under `[lints]` a particular lint belongs under, it is the part before `::` in the lint
+name.  If there isn't a `::`, then the tool is `rust`.  For example a warning
+about `unsafe` would be `lints.rust.unsafe` but a lint about
+`clippy::enum_glob_use` would be `lints.clippy.enum_glob_use`.
+
+## The `lints` table
+
+*as a new [`[workspace]` entry](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspace-section)*
+
+The `workspace.lints` table is where you define lint configuration to be inherited by members of a workspace.
+
+Specifying a workspace lint configuration is similar to package lints.
+
+Example:
+
+```toml
+# [PROJECT_DIR]/Cargo.toml
+[workspace]
+members = ["crates/*"]
+
+[workspace.lints.rust]
+unsafe = "forbid"
+```
+
+```toml
+# [PROJECT_DIR]/crates/bar/Cargo.toml
+[package]
+name = "bar"
+version = "0.1.0"
+
+[lints]
+workspace = true
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When parsing a manifest, cargo will resolve workspace inheritance for
+`lints.workspace = true` as it does with basic fields, when `workspace` is
+present, no other fields are allowed to be present.  This precludes having the
+package override the workspace on a lint-by-lint basis.
+
+cargo will contain a mapping of tool to underlying command (e.g. `rust` to
+`rustc`, `clippy` to `rustc` when clippy is the driver, `rustdoc` to
+`rustdoc`).  When running the underlying command for the specified package,
+cargo will:
+1. Transform the lints from `tool.lint = level` to `--level tool::lint`
+  - Leaving off the `tool::` when it is `rust`
+  - cargo will error if `lint` contains `::` as the first part is assumed to be
+    a tool and it should be listed in that tool's table
+2. Sort them by priority and then an unspecified order within priority that we may change in the [future](#future-possibilities).
+  - On initial release, the sort will likely be reverse alphabetical which would cause `all` to be last, making it unlikely to do what the user wants which would raise awareness of the need for setting `priority` for all groups.
+3. Pass them on the command line before other configuration like
+`RUSTFLAGS`, allowing user configuration to override package configuration.
+  - These flags will be fingerprinted so changing them will cause a rebuild only
+    for the commands where they are used.
+
+Note that this means that `[lints]` does not affect dependencies which is
+normally not an issue due to `--cap-lints` being used for dependencies.
```

Comment:
Can you clarify what this means? I see some different interpretations:

1. The `[lints]` table in Package A only applies to Package A, not its dependencies.
2. The `[lints]` table in a non-local[^1] dependency does not apply to that dependency.
3. The `[lints]` table in a non-local dependency is applied to that dependency, but usually doesn't matter because non-local dependencies have `--cap-lints`.

[^1]: Where non-local is defined by cargo's [`is_local`](https://github.com/rust-lang/cargo/blob/0ad289b264f6003a663832103e2ea810e8556dbf/src/cargo/core/compiler/unit.rs#L102-L104).

The first seems reasonable to make it clear that a `[lints]` table only applies to the package where it is defined.

The second seems a little more questionable to me. For example, a user may want to see warnings in non-local dependencies (such as with `-vv`). That may be a bit of an edge case, but something feels a little uncomfortable to me with just omitting the flags.

Reply:
Updated it to be

> Note that this means that `[lints]` is only applied to the package where its
defined and not to its dependencies, local or not.  This avoids having to unify
`[lints]` tables across local packages.  Normally, lints for non-local
dependencies won't be shown anyways because of `--cap-lints`.  As for local
dependencies, they will likely have their own `[lints]` table, most the same
one, inherited from the workspace.